src/otel/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+        }
+    }
+}
+
+
+/* function that we wrap around Tokio's runtime code. This is long lived,
+ * which means it stops only when a shutdown signal is sent to the rx
+ * channel, or we terminate the process and leave memory all over.
+ */
+#[tokio::main]
+async unsafe fn runtime(
+    log_callback: unsafe extern "C" fn(*mut i8),
+    endpoint: String,
+    proto: Protocol,
+    batch_size: f64,
+    mut rx: Receiver<SpanMessage>
+) {
+    let otlp_exporter = opentelemetry_otlp::new_exporter()
+        .http()
+        .with_endpoint(endpoint)
+        .with_protocol(proto)
+        .with_timeout(Duration::new(10, 0));
+
+    // Then pass it into pipeline builder
+    let res = opentelemetry_otlp::new_pipeline()
+        .tracing()
+        .with_trace_config(Config::default().with_resource(
+            Resource::new(vec![KeyValue::new(
+                opentelemetry_semantic_conventions::resource::SERVICE_NAME,
+                "NGINX Unit",
+            )]),
+        ))
+        .with_batch_config(
+            BatchConfigBuilder::default()
+                .with_max_export_batch_size(batch_size as _)
+                .build()
+        )
+        .with_exporter(otlp_exporter)
+        .install_batch(runtime::Tokio);
+
+
+    match res {
+        Err(e) => {
+            let msg = CString::from_vec_unchecked(e.to_string().as_bytes().to_vec());
+            log_callback(msg.into_raw() as _)
+        }
+        Ok(t) => {
+            global::set_tracer_provider(t);
+            let msg = CString::from_vec_unchecked("otel exporter has been initialised".as_bytes().to_vec());
+            log_callback(msg.into_raw() as _);
+        }
+    }
+
+    // this is the block that keeps this function running until it gets shut down.
+    // @see https://tokio.rs/tokio/tutorial/channels for the inspiration.
+    while let Some(message) = rx.recv().await {
+        match message {
+            SpanMessage::Shutdown => {
+                eprintln!("it was a shutdown");
+                break;
+            }
+            SpanMessage::Span { s: _s } => {
+                // do nothing, because the point is for this _s var to be dropped here
+                // rather than where it was sent from.
+            }
+        }
+    }
+}
+
+// it's on the caller to pass in a buf of proper length
+#[no_mangle]
+pub unsafe fn nxt_otel_copy_traceparent(buf: *mut i8, span: *const BoxedSpan) {
+    if buf.is_null() || span.is_null() {
+        return;
+    }
+
+    let traceparent = format!(
+        "00-{:032x}-{:016x}-{:02x}",
+        (*span).span_context().trace_id(), // 16 chars, 32 hex
+        (*span).span_context().span_id(),  // 8 byte, 16 hex
+        (*span).span_context().trace_flags()  // 1 char, 2 hex
+    );
+
+    assert_eq!(traceparent.len(), 55);
+
+    ptr::copy_nonoverlapping(
+        traceparent.as_bytes().as_ptr(),
+        buf as _,
+        55,
+    );
+    // set null terminator
+    *buf.add(55) = b'\0' as _;
```

Comment:
55? Now where have I seen that before?

Ah yes

```c
#define NXT_OTEL_TRACEPARENT_LEN 55                                             
```

I assume these are the same 55's?

Lets use a constant for it in the rust code also...